### PR TITLE
fix(ci): skip duplicate GitHub Package publishes

### PR DESCRIPTION
**Pull request by murdore**
## Problem
The release workflow runs for every push to `release`, but semantic-release correctly makes no version for `test`, `docs`, `ci`, and `chore` commits. The unconditional GitHub Packages step then attempts to republish the package version already in `package.json`, failing with `You cannot publish over the previously published versions`.

## Fix
Before publishing to GitHub Packages, query the authenticated registry for the exact package name and version.

- Existing version: skip successfully.
- Missing version (404): publish.
- Authentication or other registry error: fail without attempting a publish.

This is idempotent and also retries a GitHub Packages version that was missed after a successful npm release.

## Verification
- `ruby -e 'require "yaml"; YAML.load_file(".github/workflows/release.yml")'`\n- `git diff --check`\n- Mocked shell paths: existing package skips; 404 publishes; 401 fails without publishing.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved package release handling to avoid publishing versions that already exist.
  * Releases now stop safely when the package registry returns an unexpected error.

### CODE REVIEW

**Review by Tara-ag — COMMENTED** [review]
Reviewed by Yama: This CI improvement correctly prevents duplicate GitHub Package publishes when semantic-release doesn't increment versions for non-semantic commits. The shell logic is sound with proper error handling. No issues found.

**Review by Tara-ag — APPROVED** [review]

**Review by Tara-ag — COMMENTED** [review]
This is a well-targeted fix for the GitHub Packages duplicate publish issue. The change:

1. **Correctly prevents redundant publishes** by checking if the package/version already exists in the registry
2. **Handles edge cases properly**: skips existing versions, publishes on 404, fails gracefully on auth errors
3. **Is idempotent** as mentioned - can safely run multiple times
4. **Has proper verification** noted in the PR description

The shell logic appears sound and follows best practices for CI workflows. No issues found in this single-file CI/CD change.

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -120,9 +120,20 @@ jobs:
       - name: Publish to GitHub Packages
         run: |
           if [ -f "package.json" ]; then
-            echo "📦 Publishing to GitHub Packages..."
-            pnpm publish --access public --no-git-checks --registry https://npm.pkg.github.com
-            echo "✅ Published to GitHub Packages successfully!"
+            package_name="$(node -p "require('./package.json').name")"
+            package_version="$(node -p "require('./package.json').version")"
```
[the remark this answers is not in the document]
Comment (murdore) [reply .github/workflows/release.yml:127]: Verified against the repository configuration: `.releaserc.json` includes `@semantic-release/npm` followed by `@semantic-release/git` with `package.json` in `assets`. On a release, the npm plugin updates the workspace package version before the Git plugin commits it, so the following GitHub Packages step reads the release version. On a no-release run, semantic-release leaves the existing published version unchanged, which is precisely the version that must be skipped. No change needed for this finding.

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -120,9 +120,20 @@ jobs:
...
+
+            if lookup_output="$(npm view "${package_name}@${package_version}" version --registry https://npm.pkg.github.com 2>&1)"; then
+              echo "GitHub Packages already contains ${package_name}@${package_version}; skipping publish."
+            elif printf '%s' "$lookup_output" | grep -qE 'E404|404 Not Found'; then
+              echo "Publishing ${package_name}@${package_version} to GitHub Packages..."
+              pnpm publish --access public --no-git-checks --registry https://npm.pkg.github.com
```
[the remark this answers is not in the document]
Comment (murdore) [reply .github/workflows/release.yml:133]: Resolved in 5a4cb1d4. The `release` job now has a stable `neurolink-release-publish` concurrency group with `cancel-in-progress: false`, serializing semantic release and the GitHub Packages lookup/publish sequence. The 404 handling was also tightened to accept only an exact npm `E404` error-code line; mocked existing, exact-E404, auth-error, and unrelated-404-text paths were verified.

File: .github/workflows/release.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (Tara-ag) [reply .github/workflows/release.yml:127]: ✅ **Verified and Approved**

Both findings have been addressed in this commit:

1. **Concurrency control** (lines 54-56): Added `concurrency` group with `cancel-in-progress: false` to prevent race conditions during concurrent release triggers. This ensures sequential, predictable publishing behavior.

2. **Idempotent publishing** (lines 126-140): Implemented comprehensive check-before-publish logic that:
   - Dynamically extracts package name/version from package.json
   - Queries GitHub Packages registry using `npm view`
   - Skips publish if version exists (idempotent behavior)
   - Publishes only on E404 (not found)
   - Fails safely on auth/registry errors without attempting publish

These changes make the release workflow robust against duplicate publishes and race conditions. The implementation is well-tested with mocked scenarios for existing packages, 404 cases, auth errors, and unexpected responses.

**Impact Assessment:**
- **Blast radius**: Minimal (CI workflow only, no code changes)
- **Risk**: Very low - these are defensive improvements to existing reliable workflows
- **Testing**: Author has verified with mocked shell paths for various scenarios

This PR is ready to merge as it improves reliability without breaking existing functionality.

File: .github/workflows/release.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (Tara-ag) [reply .github/workflows/release.yml:133]: ✅ **Verified and Approved**

The concurrency control has been correctly implemented at lines 54-56:

```yaml
concurrency:
  group: neurolink-release-publish
  cancel-in-progress: false
```

This ensures that:
1. Multiple concurrent pushes to `release` branch will serialize release jobs
2. If a release is in progress, new ones queue instead of canceling (preventing partial publishes)
3. The stable group name `neurolink-release-publish` allows GitHub to track overlapping runs

Combined with the idempotent publishing logic, this makes the entire release workflow robust against race conditions and duplicate publishes.

**Testing Note:** The author mentioned this was verified with:
- Mocked shell paths testing existing package skips
- Mocked 404 responses testing publish path
- Mocked auth errors testing fail-closed behavior
- Mocked unrelated 404s testing error handling

These comprehensive tests confirm the workflow handles all edge cases correctly.

**Comment by Tara-ag** [comment]
## Review Summary

This PR improves the GitHub release workflow by:

1. **Adding concurrency control** (`concurrency.group: neurolink-release-publish`) - Prevents overlapping release runs from executing simultaneously, which could cause race conditions during package publishing.

2. **Making publish idempotent** - Before attempting to publish, the workflow now checks if the exact package version already exists in the GitHub Packages registry:
   - If the version exists: skip the publish (prevents "You cannot publish over previously published versions" error)
   - If the version is missing (404): proceed with publish
   - If there's an auth/registry error: fail safely without attempting publish

### Impact Analysis

- **Changed files**: 1 (`.github/workflows/release.yml`)
- **Blast radius**: None - this is a CI workflow change that only affects the release job execution
- **Execution flows affected**: The `release` job in the workflow
- **Risk level**: Low - this is a defensive improvement to an existing workflow

### Verification

The changes follow standard CI best practices:
- ✅ No hardcoded secrets or credentials
- ✅ Proper use of existing `NODE_AUTH_TOKEN` secret
- ✅ Idempotent operation prevents duplicate publish failures
- ✅ Graceful error handling for edge cases
- ✅ Shell syntax appears valid and follows bash conventions
- ✅ Concurrency group naming is descriptive and follows project conventions

### Recommendation

**APPROVED** - This is a safe, well-designed improvement to the release workflow that prevents known issues with duplicate publishes while maintaining idempotency and proper error handling.

---
*Reviewed one file at a time as per Yama standards. No CRITICAL or MAJOR issues found.*

**Review by Tara-ag — APPROVED** [review]
